org.jacoco.core/src/org/jacoco/core/internal/BytecodeVersion.java

```diff
@@ -0,0 +1,78 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal;
+
+import org.objectweb.asm.Opcodes;
+
+/**
+ * Utilities to read and modify bytecode version in bytes of class.
+ */
+public final class BytecodeVersion {
+
+	private static final int VERSION_INDEX = 6;
+
+	/**
+	 * Version of the Java 10 class file format.
+	 */
+	public static final int V10 = Opcodes.V9 + 1;
+
+	private BytecodeVersion() {
+	}
+
+	/**
+	 * Reads bytecode version from given bytes of class.
+	 * 
+	 * @param b
+	 *            bytes of class
+	 * @return version of bytecode
+	 */
+	public static int read(final byte[] b) {
+		return (short) (((b[VERSION_INDEX] & 0xFF) << 8)
+				| (b[VERSION_INDEX + 1] & 0xFF));
+	}
+
+	/**
+	 * Sets bytecode version in given bytes of class.
+	 *
+	 * @param b
+	 *            bytes of class
+	 * @param version
+	 *            version of bytecode to set
+	 */
+	public static void set(final byte[] b, final int version) {
+		b[VERSION_INDEX] = (byte) (version >>> 8);
+		b[VERSION_INDEX + 1] = (byte) version;
+	}
+
+	/**
+	 * Returns given bytes of class if version is less that {@link #V10},
+	 * otherwise copy where version set to {@link Opcodes#V9}.
+	 * 
+	 * @param version
+	 *            version of bytecode
+	 * @param source
+	 *            bytes of class
+	 * @return given bytes of class if version is less than {@link #V10},
+	 *         otherwise copy where version set to {@link Opcodes#V9}
+	 */
+	public static byte[] downgradeIfNeeded(final int version,
```

Comment:
From usage pattern it appears that a overloaded version `downgradeIfNeeded(byte[] source)` would be nice. So most call sited don't have to read version themself.

Reply:
@marchof This method must be used with caution - original version must be restored in main code in 2 out of 3 call sites (needed in `ModifiedSystemClassRuntime` and `Instrumenter`, not needed only in `Analyzer`). So by not providing overloaded version and by forcing read of version, this in my opinion forces to think twice about proper usage. Other calls are in tests and mainly only for some of them overloaded version will be useful. Thus not sure that addition of overloaded version is a good idea. WDYT?

Reply:
@Godin I was also looking for better ways to make down-/upgrade fool proof. One idea was to create a stateful sinstance of the utility:

	FutureBytecodeSupport future = new FutureBytecodeSupport();
	byte[] instrumented = future.upgrade(instrument(future.downgrade(original)));

But still you could forget to apply `upgrade()`. With Java 8 we could wrap an `Function`. For now I would simply add Javadoc to our utility how to use it properly.

Reply:
@marchof so let's keep it as is